tarutil: allow file names to be specified by regexp

cnorthwood opened the pull request:
Some features do not exist in set locations, but can be anywhere in the layer.
This allows those featurefmt to specify the filenames to be scanned by
regexp, as opposed to purely by path prefix.
If any current users of this express paths which use regexp special characters
this could be a breaking change for them (with the exception of . which will
continue to work as it matches against the literal character .), however
none of the code in this repo does that.

cnorthwood committed:
tarutil: allow file names to be specified by regexp

Some features do not exist in set locations, but can be anywhere in the layer.
This allows those featurefmt to specify the filenames to be scanned by
regexp, as opposed to purely by path prefix.
If any current users of this express paths which use regexp special characters
this could be a breaking change for them (with the exception of . which will
continue to work as it matches against the literal character .), however
none of the code in this repo does that.

fixes #456

jzelinskie commented on pkg/tarutil/tarutil.go:
```diff
@@ -78,7 +80,7 @@ func ExtractFiles(r io.Reader, filenames []string) (FilesMap, error) {
 		toBeExtracted := false
 		for _, s := range filenames {
-			if strings.HasPrefix(filename, s) {
+			if match, err := regexp.MatchString("^"+s, filename); err == nil && match {
```
This leading `^` is implicit behavior and seems unnecessary.
If we have to change other files, I don't mind.

cnorthwood replied:
The `^` does change the semantics (https://play.golang.org/p/yZl7bJLzr22) - I included it to keep the behaviour in line with the previous behaviour of only checking for the prefix.

I don't think there are any changes needed to the other files as the other file definitions were simple prefixes and didn't contain any special characters, it was more a headsup in case there were other subtle implications.

KeyboardNerd replied:
Since we're moving to use the regex, I think it makes sense to enforce all the implementations with `RequiredFile` implementation to use regex.

cnorthwood replied:
okay, I've removed the `^` and updated all users to have a regexp instead in order to maintain previous behaviour

cnorthwood committed:
tarutil: convert all filename specs to regexps

This removes the previous behaviour from tarutil to do simple prefix matching.
All places where the previous prefix-based matches were specified have been
updated to use a regexp instead, maintaining previous behaviour.

jzelinskie reviewed (approved):
LGTM! Thanks for this change.
